examples/Auction/StateMachine.hs

```diff
@@ -0,0 +1,339 @@
+module Auction.StateMachine (
+  auctionTests,
+  auctionTest,
+) where
+
+import Hedgehog (
+  Callback (Ensure, Require, Update),
+  Command (Command),
+  Concrete (Concrete),
+  FunctorB,
+  Group (Group),
+  MonadGen,
+  Property,
+  PropertyT,
+  Rec (Rec),
+  Symbolic,
+  TraversableB,
+  Var (Var),
+  annotateShow,
+  concrete,
+  executeSequential,
+  failure,
+  forAll,
+  property,
+  (===),
+ )
+import Hedgehog.Internal.Distributive (MonadTransDistributive (distributeT))
+
+import Control.Arrow (second)
+import Control.Monad (guard, liftM2)
+import Control.Monad.State.Strict (StateT, evalStateT, runState, state)
+import Data.Function ((&))
+import Data.Kind (Type)
+import Data.List (sort)
+import Data.Map (Map)
+import Data.Maybe (fromMaybe, isJust, isNothing)
+import GHC.Generics (Generic)
+import Plutus.Model (
+  Mock,
+  Run (Run),
+  adaValue,
+  checkErrors,
+  defaultBabbage,
+  initMock,
+ )
+import PlutusLedgerApi.V1 (PubKeyHash)
+import PlutusLedgerApi.V1.Contexts (TxOutRef)
+
+import Auction.PSM qualified as PSM
+import Data.Map qualified as Map
+import Hedgehog.Gen qualified as Gen
+import Hedgehog.Internal.Range qualified as Range
+
+newtype User = User {name :: String}
+  deriving stock (Eq, Ord, Show)
+
+data AuctionState v = AuctionState
+  { currentBid :: Maybe (User, User, Int, Var TxOutRef v)
+  , -- auction owner, current bidder, bid, current utxo
+    winner :: Maybe User
+  , users :: Map User (Int, Var PubKeyHash v)
```

Comment:
Why does bidding need to be restricted to specific users? Also, doesn't PSM do some of this bookkeeping?

Reply:
If you're referring to this filter:
https://github.com/mlabs-haskell/hedgehog-plutus-simple/pull/4/files#diff-3823bbcad1958e7369e3066c5f728f20d800e331deba43cdde7a857b78e18681R131-R134
I restrict bidding to users who can afford to bid because otherwise it would fail to balance ofc, and I prevent the current winner from bidding because it was causing balance issues in some cases too.

If the user has 15 but 10 is already in a bid and they try to build 12 It wouldn't balance because it tries to refund the old user 10 and pay 12 into the contract but the user and old utxo only have 15 between them. It could be fixed with a special case for when the refunded user is the same as the new bidder but it didn't seem worth it for this example.

If you're referring to the auction data type I'm not sure what you mean, the Map is not a subset of users who are allowed to bid it's just a list of all the users.

It looks like PSM can keep track of names for pub key hashes, I think I tried that briefly ran into some issue and just tracked them myself. Would you like me to change it to use PSM?